nebula_ros/test/test_hesai_functional_safety.cpp

```diff
@@ -0,0 +1,200 @@
+    {"exempt_error_leave_warn",
+     Severity::OK,
+     {0x1111, 0x2222, 0x3333},
+     {0x3333},
+     DiagnosticStatus::WARN},
+    {"exempt_all_but_ok",
+     Severity::OK,
+     {0x1111, 0x2222, 0x3333},
+     {0x2222, 0x3333},
+     DiagnosticStatus::OK},
+    // If all are exempted, the status is OK
+    {"exempt_all",
+     Severity::ERROR,
+     {0x1111, 0x2222, 0x3333},
+     {0x1111, 0x2222, 0x3333},
+     DiagnosticStatus::OK},
+  };
+}
+
+}  // namespace
+
+class CsvReaderTest : public ::testing::TestWithParam<CsvTestCase>
+{
+};
+
+TEST_P(CsvReaderTest, ReadCsv)
+{
+  const auto & test_case = GetParam();
+  if (!test_case.expected_valid) {
+    EXPECT_THROW(nebula::ros::read_error_definitions_from_csv(test_case.path), std::runtime_error);
+    return;
+  }
+
+  EXPECT_NO_THROW(nebula::ros::read_error_definitions_from_csv(test_case.path));
+
+  auto parsed = nebula::ros::read_error_definitions_from_csv(test_case.path);
+  boost::range::sort(parsed, [](const auto & a, const auto & b) { return a.code < b.code; });
+
+  auto expected = test_case.expected_error_definitions;
+  boost::range::sort(expected, [](const auto & a, const auto & b) { return a.code < b.code; });
+
+  EXPECT_EQ(parsed, expected);
+}
+
+INSTANTIATE_TEST_SUITE_P(
+  CsvReaderTest, CsvReaderTest, testing::ValuesIn(csv_test_cases()),
+  [](const testing::TestParamInfo<CsvTestCase> & p) { return p.param.path.stem().string(); });
+
+TEST(FunctionalSafetyAdvanced, ConstructorValidation)
+{
+  EXPECT_NO_THROW(nebula::ros::FunctionalSafetyAdvanced({}, {}));
+  EXPECT_NO_THROW(nebula::ros::FunctionalSafetyAdvanced({{0x1111, "Aaa", Severity::OK}}, {}));
+  EXPECT_NO_THROW(nebula::ros::FunctionalSafetyAdvanced({}, {0x1111}));
+  EXPECT_NO_THROW(nebula::ros::FunctionalSafetyAdvanced({{0x1111, "Aaa", Severity::OK}}, {0x1111}));
+}
+
+class PopulateStatusTest : public ::testing::TestWithParam<PopulateStatusTestCase>
+{
+};
+
+TEST_P(PopulateStatusTest, PopulateStatus)
+{
+  const auto & test_case = GetParam();
+  nebula::ros::FunctionalSafetyAdvanced functional_safety_advanced(
+    error_definitions(), test_case.exempted_error_codes);
+  diagnostic_msgs::msg::DiagnosticStatus status;
+  functional_safety_advanced.populate_status(
+    test_case.arg_severity, test_case.arg_error_codes, status);
+  EXPECT_EQ(status.level, test_case.expected_status_level);
+}
+
+INSTANTIATE_TEST_SUITE_P(
+  PopulateStatusTest, PopulateStatusTest, testing::ValuesIn(populate_status_test_cases()),
+  [](const testing::TestParamInfo<PopulateStatusTestCase> & p) { return p.param.name; });
+
+TEST(FunctionalSafetyAdvanced, PopulateStatus)
+{
+  nebula::ros::FunctionalSafetyAdvanced functional_safety_advanced(error_definitions(), {});
+  diagnostic_msgs::msg::DiagnosticStatus status;
+  functional_safety_advanced.populate_status(Severity::OK, {0x1111}, status);
+  EXPECT_EQ(status.level, diagnostic_msgs::msg::DiagnosticStatus::OK);
+  EXPECT_EQ(status.message, "Operating nominally");
```

Comment:
Operating ... normally?

Reply:
That name is from #324 - is it not correct English? I though it sounded cool :bow: